core/services/log/relayer.go

```diff
@@ -0,0 +1,350 @@
+package log
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+type relayer struct {
+	orm       ORM
+	listeners map[common.Address]map[Listener]struct{}
+
+	latestBlock                  uint64
+	fewestRequestedConfirmations uint64
+
+	addListener      *utils.Mailbox
+	rmListener       *utils.Mailbox
+	newHeads         *utils.Mailbox
+	newLogs          *utils.Mailbox
+	connectionEvents *utils.Mailbox
+
+	utils.StartStopOnce
+	utils.DependentAwaiter
+	chStop chan struct{}
+	chDone chan struct{}
+}
+
+// A `registration` represents a Listener's subscription to the logs of a
+// particular contract.
+type registration struct {
+	address  common.Address
+	listener Listener
+}
+
+type connectionEvent int
+
+const (
+	disconnected connectionEvent = iota
+	connected
+)
+
+func newRelayer(orm ORM, dependentAwaiter utils.DependentAwaiter) *relayer {
+	return &relayer{
+		orm:              orm,
+		listeners:        make(map[common.Address]map[Listener]struct{}),
+		addListener:      utils.NewMailbox(0),
```

Comment:
Wouldn't a regular channel work here?

Reply:
Don't want it to block, but also don't want to drop subscriptions